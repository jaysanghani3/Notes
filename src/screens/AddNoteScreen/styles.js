import {StyleSheet} from 'react-native';
import Colors from '../../utils/Colors';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 14,
    justifyContent: 'space-between',
  },
  row: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  title: {
    fontSize: 25,
    fontWeight: 'bold',
    padding: 5,
    borderBottomWidth: 1,
    borderColor: Colors.gray,
  },
  detailsInput: {
    fontSize: 17,
    padding: 10,
    borderWidth: 1,
    borderColor: Colors.gray,
    borderRadius: 10,
    marginVertical: 10,
    maxHeight: 250,
  },
  btnGrp: {
    flexDirection: 'row',
    justifyContent: 'space-between',
  },
  btn: {
    backgroundColor: Colors.violet,
    padding: 15,
    borderRadius: 8,
    marginBottom: 5,
    width: '48%',
  },
  btnFont: {
    textAlign: 'center',
    color: Colors.white,
    fontWeight: 'bold',
  },
  noteView: {
    flex: 1,
    backgroundColor: Colors.white,
    padding: 20,
    borderRadius: 8,
  },
  noteViewTitle: {
    fontSize: 35,
    fontWeight: 'bold',
    color: Colors.violet,
    borderBottomColor: Colors.gray,
    borderBottomWidth: 1,
    paddingBottom: 10,
  },
  noteViewDetails: {
    fontSize: 16,
    color: Colors.gray,
    marginTop: 30,
  },
  titleCount: {
    textAlign: 'right',
    color: Colors.gray,
  },
  maxCount: {
    textAlign: 'right',
    color: Colors.error,
  },
});

export default styles;
