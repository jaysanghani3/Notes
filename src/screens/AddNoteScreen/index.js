import {
  ScrollView,
  Text,
  TextInput,
  TouchableOpacity,
  View,
  Keyboard,
} from 'react-native';
import React, {useEffect, useState} from 'react';
import styles from './styles';
import NetInfo from '@react-native-community/netinfo';
import Loader from '../../components/Loader';
import globalStyles from '../../utils/globalStyles';
import Strings from '../../translation/Strings';
import {
  addNotes,
  updateNote,
  updateNoteOffline,
  addNoteOffline,
} from '../../services/firestore';

const AddNoteScreen = ({route, navigation}) => {
  const {noteTitle, noteDetails, noteId, screen} = route.params;
  const [title, setTitle] = useState(noteTitle);
  const [details, setDetails] = useState(noteDetails);
  const [isLoading, setIsLoading] = useState(false);
  const [isConnected, setIsConnected] = useState(true);
  const [btnText, setBtnText] = useState(Strings.addBtn);
  const [titleLength, setTitleLength] = useState(title.length);
  const [titleError, setTitleError] = useState('');

  useEffect(() => {
    const unsubscribe = NetInfo.addEventListener(state => {
      setIsConnected(state.isConnected);
    });
    return () => {
      unsubscribe();
    };
  }, []);

  useEffect(() => {
    if (screen === Strings.addScreenTitle) {
      navigation.setOptions({title: Strings.addScreenTitle});
      setBtnText(Strings.addBtn);
    } else if (screen === Strings.editScreenTitle) {
      navigation.setOptions({title: Strings.editScreenTitle});
      setBtnText(Strings.updateBtn);
    } else if (screen === Strings.viewScreenTitle) {
      navigation.setOptions({title: Strings.viewScreenTitle});
    }
  }, [navigation]);

  const handleOfflineSupport = async () => {
    try {
      navigation.goBack();
      if (screen === Strings.editScreenTitle) {
        updateNoteOffline(noteId, title, details);
      } else if (screen === Strings.addScreenTitle) {
        addNoteOffline(title, details);
      }
      setIsLoading(false);
    } catch (e) {
      console.log(e);
    }
  };

  const handleSave = async () => {
    Keyboard.dismiss();
    if (!title && !details) {
      setTitleError(Strings.required);
    } else if (!title) {
      setTitleError(Strings.required);
    } else if (!details) {
    } else {
      try {
        if (!isConnected) {
          handleOfflineSupport();
        } else {
          const newNoteId = Math.random().toString(36).substring(7);
          if (screen === Strings.editScreenTitle) {
            updateNote(noteId, title, details);
          } else if (screen === Strings.addScreenTitle) {
            addNotes(newNoteId, title, details);
          }
          navigation.goBack();
        }
      } catch (e) {
        console.log(e);
      }
    }
  };

  const handleTitleChange = title => {
    setTitle(title);
    setTitleLength(title.length);
    title.length >= Strings.titleLength
      ? setTitleError(Strings.titleLengthError)
      : setTitleError('');
  };

  const handleDetailsChange = details => {
    setDetails(details);
  };

  return (
    <>
      {screen === Strings.viewScreenTitle ? (
        <View style={styles.noteView}>
          <Text style={styles.noteViewTitle}>{noteTitle}</Text>
          <ScrollView>
            <Text style={styles.noteViewDetails}>{noteDetails}</Text>
          </ScrollView>
        </View>
      ) : (
        <View style={styles.container}>
          <Loader state={isLoading} />

          <View>
            <TextInput
              placeholder={Strings.titlePlaceholder}
              maxLength={Strings.titleLength}
              style={styles.title}
              defaultValue={title}
              onChangeText={title => {
                handleTitleChange(title);
              }}
            />
            <View style={styles.row}>
              <Text style={globalStyles.error}>{titleError}</Text>
              <Text
                style={
                  titleLength === Strings.titleLength
                    ? styles.maxCount
                    : styles.titleCount
                }>
                {titleLength}/{Strings.titleLength}
              </Text>
            </View>
            <ScrollView>
              <TextInput
                placeholder={Strings.detailsPlaceholder}
                style={styles.detailsInput}
                multiline={true}
                numberOfLines={10}
                textAlignVertical="top"
                defaultValue={details}
                onChangeText={details => {
                  handleDetailsChange(details);
                }}
              />
            </ScrollView>
          </View>
          <View style={styles.btnGrp}>
            <TouchableOpacity
              style={styles.btn}
              onPress={() => {
                setDetails('');
                setTitle('');
              }}>
              <Text style={styles.btnFont}>{Strings.resetBtn}</Text>
            </TouchableOpacity>
            <TouchableOpacity style={styles.btn} onPress={handleSave}>
              <Text style={styles.btnFont}>{btnText}</Text>
            </TouchableOpacity>
          </View>
        </View>
      )}
    </>
  );
};
export default AddNoteScreen;
