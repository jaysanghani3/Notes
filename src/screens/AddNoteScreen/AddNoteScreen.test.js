import React from 'react';
import renderer from 'react-test-renderer';
import {test, expect} from '@jest/globals';
import AddNoteScreen from './index';

test('renders correctly', () => {
  const mockRoute = {
    params: {
      screen: 'test',
      noteTitle: 'test',
      noteDetails: 'test',
      noteId: 'test',
    },
  };

  const tree = renderer
    .create(
      <AddNoteScreen navigation={{navigate: jest.fn()}} route={mockRoute} />,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
