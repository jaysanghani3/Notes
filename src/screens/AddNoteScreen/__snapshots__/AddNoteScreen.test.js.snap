// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    {
      "flex": 1,
      "justifyContent": "space-between",
      "padding": 14,
    }
  }
>
  <View>
    <TextInput
      defaultValue="test"
      maxLength={15}
      onChangeText={[Function]}
      placeholder="Title"
      style={
        {
          "borderBottomWidth": 1,
          "borderColor": "#B6B6B6",
          "fontSize": 25,
          "fontWeight": "bold",
          "padding": 5,
        }
      }
    />
    <View
      style={
        {
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <Text
        style={
          {
            "color": "#D32F2F",
          }
        }
      />
      <Text
        style={
          {
            "color": "#B6B6B6",
            "textAlign": "right",
          }
        }
      >
        4
        /
        15
      </Text>
    </View>
    <RCTScrollView>
      <View>
        <TextInput
          defaultValue="test"
          multiline={true}
          numberOfLines={10}
          onChangeText={[Function]}
          placeholder="Write your note details here..."
          style={
            {
              "borderColor": "#B6B6B6",
              "borderRadius": 10,
              "borderWidth": 1,
              "fontSize": 17,
              "marginVertical": 10,
              "maxHeight": 250,
              "padding": 10,
            }
          }
          textAlignVertical="top"
        />
      </View>
    </RCTScrollView>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
        "justifyContent": "space-between",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#6574FF",
          "borderRadius": 8,
          "marginBottom": 5,
          "opacity": 1,
          "padding": 15,
          "width": "48%",
        }
      }
    >
      <Text
        style={
          {
            "color": "#ffffff",
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        Reset
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#6574FF",
          "borderRadius": 8,
          "marginBottom": 5,
          "opacity": 1,
          "padding": 15,
          "width": "48%",
        }
      }
    >
      <Text
        style={
          {
            "color": "#ffffff",
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        Add
      </Text>
    </View>
  </View>
</View>
`;
